bldsys/cmake/shader_helper.cmake

```diff
@@ -0,0 +1,123 @@
+add_custom_target(vkb__shaders)
+
+# TODO: Shaders to fix
+set(SHADERS_TO_FIX
+    buffer_array.frag
+    buffer_array.vert
+)
+
+macro(get_hlsl_target_profile SHADER_FILE OUT_PROFILE)
+    if(${SHADER_FILE} MATCHES ".*\\.vert$")
+        set(${OUT_PROFILE} "vs_6_0")
+    elseif(${SHADER_FILE} MATCHES ".*\\.frag$")
+        set(${OUT_PROFILE} "ps_6_0")
+    else()
+        message(FATAL_ERROR "Unknown shader profile for ${SHADER_FILE}")
+    endif()
+endmacro()
+
+function(compile_shaders)
+    set(options)  
+    set(oneValueArgs ID SPIRV_VERSION)
+    set(multiValueArgs SHADER_FILES_GLSL SHADER_FILES_HLSL)
+
+    cmake_parse_arguments(TARGET "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if (NOT TARGET_SPIRV_VERSION)
+        set(TARGET_SPIRV_VERSION "1.0")
+    endif()
+
+    find_program(SPV_VALIDATOR_EXECUTABLE spirv-val HINTS $ENV{VULKAN_SDK}/bin)
+    if(NOT SPV_VALIDATOR_EXECUTABLE)
+        message(WARNING "spirv-val not found! SPIR-V validation will not be performed. Please set VULKAN_SDK to the Vulkan SDK path.")
+    endif()
+
+    find_program(GLSLC_EXECUTABLE glslc HINTS $ENV{VULKAN_SDK}/bin)
+    if(NOT GLSLC_EXECUTABLE)
+        message(FATAL_ERROR "glslc not found! Shaders with .glsl extension will not be compiled to SPIR-V. Please set VULKAN_SDK to the Vulkan SDK path.")
+    endif()
```

Comment:
So, this won't work that well for us, we don't generally have the Vulkan-SDK on our build servers.
Could we not just add glslangValidator and spirv-tools as submodules and build these as required?

Reply:
If we bundle the SPIRV with the repository instead of ignoring it then there wouldn't be a need to install any tools out of the box? Tools would only be needed if you change the shaders

Is there any capability for you to install tools on your build servers before running a build task? Possibly with a script?

Reply:
> If we bundle the SPIRV with the repository instead of ignoring it then there wouldn't be a need to install any tools out of the box?

If they're bundled that's fine.

> Is there any capability for you to install tools on your build servers before running a build task?

Anything is possible, I'm just looking for the path of least resistance.

Reply:
Bundling is likely best here. Only maintainers of the shaders would require the tools needed to compile the shaders

There are only a few shaders which have variants (<5) I'm hoping to be able to simplify them further to remove the need entirely. That would mean that we would only have a single SPIRV per shader

Reply:
Wait, does that imply that building SPIR-V at build time when Sanders change locally is no longer a feature of this design?  Right now if someone changes a shader and reruns an app they see the results of the change.  Losing that seems like it would be a big deal.

Or is the idea that the bundled code only gets used as a fallback?

Reply:
It just means that if you do want to alter the shaders you should at least have the tools installed to offline compile shaders. Most users should have this already with the Vulkan SDK

From the discussions in the working group the rough requirements are:
- Include HLSL alternatives for all shaders in the long run
- Do not include DXC or shader compilers within the project as these are assumed to be installed locally
- Prefer offline compilation over online compilation due to the point above
- Keep things simple with a single compile pass and document this for users

Previous attempts at solving this have been deemed too complex, hence this approach is all about being simple and removing complexity / unneeded shader usage.

Just updated the todo, I have a few samples left to simplify. Most samples work out of the box with this simpler model

Current bench mark for recompiling all shaders

`cmake --build build --target vkb__shaders --parallel  0.98s user 0.91s system 456% cpu 0.416 total`